internal/lsp/server.go

```diff
@@ -151,105 +172,199 @@ func (s *Server) UnwatchFiles(handle project.WatcherHandle) error {
 }
 
 // RefreshDiagnostics implements project.Client.
-func (s *Server) RefreshDiagnostics() error {
+func (s *Server) RefreshDiagnostics(ctx context.Context) error {
 	if ptrIsTrue(s.initializeParams.Capabilities.Workspace.Diagnostics.RefreshSupport) {
-		if err := s.sendRequest(lsproto.MethodWorkspaceDiagnosticRefresh, nil); err != nil {
+		if _, err := s.sendRequest(ctx, lsproto.MethodWorkspaceDiagnosticRefresh, nil); err != nil {
 			return fmt.Errorf("failed to refresh diagnostics: %w", err)
 		}
 	}
 	return nil
 }
 
 func (s *Server) Run() error {
+	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer stop()
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error { return s.dispatchLoop(ctx) })
+	g.Go(func() error { return s.writeLoop(ctx) })
+	g.Go(func() error { return s.readLoop(ctx) })
+	return g.Wait()
+}
+
+func (s *Server) readLoop(ctx context.Context) error {
 	for {
-		req, err := s.read()
+		msg, err := s.read()
 		if err != nil {
+			if err == io.EOF {
```

Comment:
I wonder when this regressed; errorlint used to complain about it, then it was fixed, and I guess now it's broken again?

Reply:
Wait, the linter _is_ complaining about this; I thought you told me to change it to be this way?

Reply:
I did, but I guess that linter was not fully fixed to stop complaining about this ☹️

I guess feel free to `errors.Is` it and we can come back to this later.